### PR TITLE
MCO-654: forcefile should always trigger an OS update

### PR DESCRIPTION
**Pull request by cdoern**
**- What I did**

currently, the forcefile sometimes gets stopped from triggering an osupdate if one of our other priorities gets in the way. The forcefile should alwyas trigger and end to end update, re-applying all changes if the osImageURL is different. This will get many customers out of this half baked reboot loop we have been seeing.

**- How to verify it**

when using the forcefile, should always see `Updating OS` indicating a full reboot, especially when there is a config mismatch with the osImageURLs.

**- Description for the changelog**

the forcefile now always triggers a full os update.

### CODE REVIEW

**Comment by djoshy** [comment]
Looks like a forcefile test failure, might need some refactor since we are changing the behavior

**Comment by djoshy** [comment]
Also, this might need a rebase once https://github.com/openshift/machine-config-operator/pull/3583 lands

File: pkg/daemon/update.go [hunk 1]
```diff
@@ -256,7 +256,7 @@ func removePendingDeployment() error {
 }
 
 // applyOSChanges extracts the OS image and adds coreos-extensions repo if we have either OS update or package layering to perform
-func (dn *CoreOSDaemon) applyOSChanges(mcDiff machineConfigDiff, oldConfig, newConfig *mcfgv1.MachineConfig) (retErr error) {
+func (dn *CoreOSDaemon) applyOSChanges(mcDiff machineConfigDiff, oldConfig, newConfig *mcfgv1.MachineConfig, force bool) (retErr error) {
 	// We previously did not emit this event when kargs changed, so we still don't
 	if mcDiff.osUpdate || mcDiff.extensions || mcDiff.kernelType {
```
Comment (yuqi-zhang) [inline pkg/daemon/update.go:261]: I think you need to also check for the force here

File: pkg/daemon/update.go [hunk 2]
```diff
@@ -1918,12 +1923,31 @@ func (dn *Daemon) reboot(rationale string) error {
...
+}
+
+func (dn *CoreOSDaemon) applyLayeredOSChanges(mcDiff machineConfigDiff, oldConfig, newConfig *mcfgv1.MachineConfig, force bool) (retErr error) {
 	// Override the computed diff if the booted state differs from the oldConfig
 	// https://issues.redhat.com/browse/OCPBUGS-2757
 	if mcDiff.osUpdate && dn.bootedOSImageURL == newConfig.Spec.OSImageURL {
-		klog.Infof("Already in desired image %s", newConfig.Spec.OSImageURL)
-		mcDiff.osUpdate = false
+		if force {
+			if err := dn.changeBootedImage(""); err != nil {
```
Comment (yuqi-zhang) [inline pkg/daemon/update.go:1944]: I don't think this is necessary if we're forcing it through to here. I think it's also safe to remove the mcDiff.osUpdate check from above and continue with the rest via force. Even if we force, I think if dn.bootedOSImageURL == newConfig.Spec.OSImageURL means that we are on the right version

**Comment by cdoern** [comment]
Testing this with the suite is to costly. I ran tests by hand.

1) created a cluster with this PR.
2) found a new osImageURL from a nightly payload (rpm-ostree is the name)
3) ran rpm-ostree rebase --experimental ostree-unverified-registry:<URL>
4) waited for the daemon to mark the node as degraded, I applied the forcefile
5) node rebooted, marked as done with new osImageURL

File: test/e2e-shared-tests/mcd_config_drift.go [hunk 3]
```diff
@@ -237,6 +237,7 @@ func (c configDriftTest) Run(t *testing.T) {
 					t.Logf("Setting forcefile to initiate recovery (%s)", constants.MachineConfigDaemonForceFile)
 					helpers.ExecCmdOnNode(t, c.ClientSet, c.node, "touch", filepath.Join("/rootfs", constants.MachineConfigDaemonForceFile))
 				})
+
```
Comment (sinnykumari) [inline test/e2e-shared-tests/mcd_config_drift.go:240]: nit: unrelated change

**Comment by sinnykumari** [comment]
Also, we may want to get https://docs.openshift.com/container-platform/4.13/post_installation_configuration/machine-configuration-tasks.html updated for force file related change.

File: pkg/daemon/update.go [hunk 4]
```diff
@@ -680,8 +681,12 @@ func newMachineConfigDiff(oldConfig, newConfig *mcfgv1.MachineConfig) (*machineC
 	kargsEmpty := len(oldConfig.Spec.KernelArguments) == 0 && len(newConfig.Spec.KernelArguments) == 0
 	extensionsEmpty := len(oldConfig.Spec.Extensions) == 0 && len(newConfig.Spec.Extensions) == 0
 
+	force := false
+	if _, err := os.Stat(constants.MachineConfigDaemonForceFile); err == nil {
+		force = true
+	}
 	return &machineConfigDiff{
-		osUpdate:   oldConfig.Spec.OSImageURL != newConfig.Spec.OSImageURL,
+		osUpdate:   oldConfig.Spec.OSImageURL != newConfig.Spec.OSImageURL || force,
```
[the remark this answers is not in the document]
Comment (cdoern) [reply pkg/daemon/update.go:689]: well, that osUpdate will not happen unless the osImageURL is different, which in that case we should execute a forced update.

**Comment by cdoern** [comment]
I do not think this needs QE testing as it is nothigh risk for any current functionality. @yuqi-zhang wdyt?
